plutus-emulator/src/Wallet/Emulator/NodeClient.hs

```diff
@@ -0,0 +1,136 @@
+{-# LANGUAGE ConstraintKinds       #-}
+{-# LANGUAGE DeriveAnyClass        #-}
+{-# LANGUAGE DerivingStrategies    #-}
+{-# LANGUAGE FlexibleContexts      #-}
+{-# LANGUAGE FlexibleInstances     #-}
+{-# LANGUAGE GADTs                 #-}
+{-# LANGUAGE LambdaCase            #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE OverloadedStrings     #-}
+{-# LANGUAGE TemplateHaskell       #-}
+{-# LANGUAGE TupleSections         #-}
+{-# LANGUAGE TypeApplications      #-}
+module Wallet.Emulator.NodeClient where
+
+import           Control.Lens        hiding (index)
+import           Control.Monad.State
+import           Data.Aeson          (FromJSON, ToJSON)
+import           Data.List           (partition)
+import           Data.Maybe          (isNothing)
+import           Data.Traversable    (for)
+import           GHC.Generics        (Generic)
+
+import           Ledger              (Blockchain, Slot (..), Tx (..), TxId, txId)
+import qualified Ledger.Index        as Index
+import qualified Ledger.Interval     as Interval
+
+
+-- | Events produced by the blockchain emulator.
+data ChainEvent =
+    TxnSubmit TxId
+    -- ^ A transaction has been added to the pool of pending transactions.
+    | TxnValidate TxId
+    -- ^ A transaction has been validated and added to the blockchain.
+    | TxnValidationFail TxId Index.ValidationError
+    -- ^ A transaction failed  to validate.
+    | SlotAdd Slot
```

Comment:
Looking at this, it feels like these aren't all at the same level. Transactions validating and failing to validate can't happen on their own. Perhaps there should be a `BlockAdd` event which has a list of `TxnValidate`/`TxnValidationFail`s inside it?

But I don't think it matters specifically for now.

Reply:
Isn't that just `ValidatedBlock` ?

Reply:
Yes. There is a similar problem with Types/Notification, where you have a block add / slot change event. They always happen together AFAIK.

Reply:
I think this will become clearer as we disentangle the pieces.